### PR TITLE
Add gcf functions to trigger bt_setup and airflow trigger.

### PR DESCRIPTION
**Pull request by sushant2603**
This add two GCF functions

1) triggers at the end of flume run and setups BT and writes to GCS to trigger airflow job.
2) Configures and triggers airflow job.

In future we can merge these both.

Currently working on tests and documentation for all the steps.

### CODE REVIEW

**Comment by sushant2603** [comment]
@pradh : Can you take  look at this. no rush.

File: cloud_automation/gcf/airflow_trigger.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
+""" This module triggers when a file is writtent to prophet-cache GCS bucket.
```
Comment (pradh) [inline cloud_automation/gcf/airflow_trigger.py:1]: written

File: cloud_automation/gcf/airflow_trigger.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    webserver_id = 'x7fea1724da0f8c91p-tp'
+    # The name of the DAG is configured as part of the airflow job.
+    dag_name = 'GcsToBTCache'
+    webserver_url = (
+        'https://'
+        + webserver_id
+        + '.appspot.com/api/experimental/dags/'
+        + dag_name
+        + '/dag_runs'
+    )
```
Comment (pradh) [inline cloud_automation/gcf/airflow_trigger.py:71]: For my understanding, who sets up this URL?

File: cloud_automation/gcf/airflow_trigger.py (hunk 1 above)
Comment (pradh) [reply cloud_automation/gcf/airflow_trigger.py:1]: written to prophet-cache GCS bucket by <what> running in borg (?).

File: cloud_automation/gcf/airflow_trigger.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+  name = data['name']
+
+  # Check if this is triggered after flume job completion
+  if name.endswith(PIPELINE_TRIGGER_FILE):
+    csv_file = read_contents(bucket, name)
+    path = pathlib.PurePath(csv_file)
+    # path.parent.name gives the last directory in the path. We use this as
+    # bt table id.
+    bt_table_id = path.parent.name
+    trigger_dag(bt_table_id, csv_file)
```
Comment (pradh) [inline cloud_automation/gcf/airflow_trigger.py:43]: This is the cache_csv_file ? (just to be clear)

File: cloud_automation/gcf/airflow_trigger.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+FAILURE_FILE = 'failure.txt'
+
+def read_contents(bucket, name):
+  client = storage.Client()
+  blob = client.get_bucket(bucket).blob(name)
+  result = blob.download_as_string()
+  return result
+
+
+def gcs_trigger(data, context=None):
```
Comment (pradh) [inline cloud_automation/gcf/airflow_trigger.py:32]: If this function errors out, where will it notify? Could you add some notes about it to http://go/dccachedocs.

File: cloud_automation/gcf/bt_trigger.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
+package btcachegeneration
```
Comment (pradh) [inline cloud_automation/gcf/bt_trigger.go:1]: Please add some comments on what this does, and importantly, when in the flow this gets triggered and by whom.

File: cloud_automation/gcf/bt_trigger.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+		log.Printf("Unable to increase bigtable cluster size: %v", err)
+		return err
+	}
+	return nil
+}
+
+// GCSTrigger consumes a GCS event.
+func GCSTrigger(ctx context.Context, e GCSEvent) error {
+	return nil
+	// Read contents of GCS file.
```
Comment (pradh) [inline cloud_automation/gcf/bt_trigger.go:96]: Is this the same GCS file containing the cache CSV file?

File: cloud_automation/gcf/bt_trigger.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	rc, err := bucket.Object(fileName).NewReader(ctx)
+	if err != nil {
+		log.Printf("Unable to open file from bucket %q, file %q: %v\n", bucketName, fileName, err)
+		return nil, fmt.Errorf("Unable to open file from bucket %q, file %q: %v", bucketName, fileName, err)
+	}
+	defer rc.Close()
+	return ioutil.ReadAll(rc)
+}
+
+func SetupBigtable(ctx context.Context, tableID string) error {
```
Comment (pradh) [inline cloud_automation/gcf/bt_trigger.go:49]: nit: Add function comments to say it creates table, cluster and scales up nodes?

File: cloud_automation/gcf/bt_trigger.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+	}
+
+	// Scale up bigtable cluster. This helps speed up the dataflow job.
+	// We scale down again once dataflow job completes.
+	instanceAdminClient, err := bigtable.NewInstanceAdminClient(ctx, projectID)
+	if err != nil {
+		log.Printf("Unable to create a table instance admin client. %v", err)
+		return err
+	}
+	if err := instanceAdminClient.UpdateCluster(dctx, bigtableInstance, bigtableCluster, bigtableNodes); err != nil {
```
Comment (pradh) [inline cloud_automation/gcf/bt_trigger.go:86]: For my understanding, where does scaling down happen?

File: cloud_automation/gcf/bt_trigger.go [hunk 9]
```diff
@@ -0,0 +1,153 @@
+// Package btcachegeneration runs a GCF function that triggers in 2 scenarios:
+// 1) completion of prophet-flume job in borg. 
+//    The trigger is based on GCS file prophet-cache/latest_base_cache_run.txt.
```
Comment (pradh) [inline cloud_automation/gcf/bt_trigger.go:3]: This has changed since?

**Comment by sushant2603** [comment]
@pradh This is synced to latest version now.

**Comment by shifucun** [comment]
@pradh is this PR still needed or can close it?

**Comment by pradh** [comment]
I think this PR is still relevant and deployed.

@sushant2603 - could you confirm?

If so, I'll merge and then take care of any required follow ups.

**Comment by pradh** [comment]
I'm going to merge this and perform any follow up changes.

**Pull request merged**